### PR TITLE
EnC - Tell the debugger about updated type def tokens

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/52585

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 1]
```diff
@@ -2739,6 +2739,28 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
...
+        }
+
+        /// <summary>
+        /// Emit the differences between the compilation and the previous generation
+        /// for Edit and Continue. The differences are expressed as added and changed
+        /// symbols, and are emitted as metadata, IL, and PDB deltas. A representation
+        /// of the current compilation is returned as an EmitBaseline for use in a
+        /// subsequent Edit and Continue.
+        /// </summary>
+        public EmitDifferenceResult EmitDifference(
```
Comment (davidwengier) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:2754]: This upsets the RS0026 analyzer, so the build is failing, but I cannot work out what to do to fix it. It says not to add multiple overloads with optional parameters, but there are already multiple overloads with optional parameters so any of the proposed solutions would _break_ compatibility. Plus, I'm not adding an optional parameter ¯\\\_(ツ)_/¯

Comment (mavasani) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:2754]: Suppress it. RS0026 needs to be tweaked to handle more cases. We have multiple suppressions for it throughout Roslyn repo.

Comment (davidwengier) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:2754]: Thanks! Almost too easy :)

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 2]
```diff
@@ -2739,6 +2740,28 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
...
+            EmitBaseline baseline,
+            IEnumerable<SemanticEdit> edits,
+            Func<ISymbol, bool> isAddedSymbol,
+            Stream metadataStream,
+            Stream ilStream,
+            Stream pdbStream,
+            ICollection<MethodDefinitionHandle> updatedMethods,
+            ICollection<TypeDefinitionHandle> updatedTypes,
```
Comment (tmat) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:2763]: >updatedTypes [](http://example.com/codeflow?start=46&length=12)

I wonder if we should move `updatedMethods` and `updatedTypes` to `EmitDifferenceResult` as `ImmutableArray<MethodDefinitionHandle>` and `ImmutableArray<MethodDefinitionHandle>`, respectively, instead.

They are outputs/results. The previous overload could also populate the collection for backward compat but we can make it [Obsolete].

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 3]
```diff
@@ -2784,6 +2808,7 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
             Stream ilStream,
             Stream pdbStream,
             ICollection<MethodDefinitionHandle> updatedMethodHandles,
+            ICollection<TypeDefinitionHandle> updatedTypeHandlers,
```
Comment (tmat) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:2811]: >r [](http://example.com/codeflow?start=63&length=1)

typo

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs [hunk 4]
```diff
@@ -228,6 +230,17 @@ public void GetMethodTokens(ICollection<MethodDefinitionHandle> methods)
             }
         }
 
+        /// <summary>
+        /// Return tokens for all modified debuggable types.
+        /// </summary>
+        public void GetTypeTokens(ICollection<TypeDefinitionHandle> types)
```
Comment (tmat) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs:236]: >GetTypeTokens [](http://example.com/codeflow?start=20&length=13)

perhaps rename to `GetUpdatedTypeHandles` - it's not immediately clear what types does this return. (same for GetMethodTokens)

**Comment by tmat** [comment]
>     public class EditAndContinueTests : EditAndContinueTestBase

Do we have tests that verify updated types for synthesized types? Local function/lambda's closures, state machines, the `<Module>` type, anonymous types, etc.

---
Refers to: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs:29 in d37f7cc. [](commit_id = d37f7cce7f3dbe1038c8f9a8991cc916dc008722, deletion_comment = False)

**Comment by tmat** [comment]
Let's target main. In general we shouldn't make changes to -vs-deps unless absolutely necessary since it just makes harder to work in main.

**Comment by davidwengier** [comment]
> In general we shouldn't make changes to -vs-deps unless absolutely necessary

The debugger API that this needs isn't in main yet, which is why I targeted 16.10-vs-deps. I'm happy to not take this for 16.1x, but at the moment we still need to target main-vs-deps for the debugger contracts library.

**Comment by tmat** [comment]
The changes in the PR do not depend on the debugger API. Passing it to the debugger can nly be done in -vs-deps, but all other changes can be in main.

**Comment by davidwengier** [comment]
Okay, I've retargeted to main.

> Do we have tests that verify updated types for synthesized types? Local function/lambda's closures, state machines, the `<Module>` type, anonymous types, etc. Also a test with updated generic type.

I've added verification to existing tests that cover those scenarios. Couldn't get `<Module>` working, looks like compiler changes would be needed to support that. Logged https://github.com/dotnet/roslyn/issues/53310

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 5]
```diff
@@ -883,6 +883,7 @@ public async Task<SolutionUpdate> EmitSolutionUpdateAsync(Solution solution, Sol
                     using var ilStream = SerializableBytes.CreateWritableStream();
 
                     var updatedMethods = ImmutableArray.CreateBuilder<MethodDefinitionHandle>();
+                    var updatedTypes = ImmutableArray.CreateBuilder<TypeDefinitionHandle>();
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs:886]: >updatedTypes [](http://example.com/codeflow?start=24&length=12)

Not needed anymore

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 6]
```diff
@@ -894,14 +895,14 @@ public async Task<SolutionUpdate> EmitSolutionUpdateAsync(Solution solution, Sol
...
-                        updatedMethods,
                         cancellationToken);
 
                     if (emitResult.Success)
                     {
                         Contract.ThrowIfNull(emitResult.Baseline);
 
-                        var updatedMethodTokens = updatedMethods.SelectAsArray(h => MetadataTokens.GetToken(h));
+                        var updatedMethodTokens = emitResult.UpdatedMethods.SelectAsArray(h => MetadataTokens.GetToken(h));
+                        var updatedTypeTokens = emitResult.UpdatedTypes.SelectAsArray(h => MetadataTokens.GetToken(h));
```
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs:905]: Perhaps add TODO that this is gonna be passed to ManagedModuleUpdate in vs-deps

**Review by tmat — APPROVED** [review]
:shipit:

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 7]
```diff
@@ -2422,6 +2435,9 @@ void M()
             using var md1 = diff1.GetMetadata();
             var reader1 = md1.Reader;
             var readers = new[] { reader0, reader1 };
+
+            diff1.VerifyUpdatedTypes("0x02000003");
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs:2439]: >0x02000003 [](http://example.com/codeflow?start=38&length=10)

I wonder if we could pass the `readers` to VerifyUpdatedTypes and resolve the tokens to type names for better readability of the tests. Something to perhaps follow up on.

**Comment by davidwengier** [comment]
@dotnet/roslyn-compiler can I get a couple of reviews?

File: src/Compilers/Test/Core/Compilation/CompilationDifference.cs [hunk 8]
```diff
@@ -150,7 +147,14 @@ public void VerifyUpdatedMethods(params string[] expectedMethodTokens)
         {
             AssertEx.Equal(
                 expectedMethodTokens,
-                UpdatedMethods.Select(methodHandle => $"0x{MetadataTokens.GetToken(methodHandle):X8}"));
+                EmitResult.UpdatedMethods.Select(methodHandle => $"0x{MetadataTokens.GetToken(methodHandle):X8}"));
+        }
+
+        public void VerifyUpdatedTypes(params string[] expecedTypeTokens)
```
Comment (RikkiGibson) [inline src/Compilers/Test/Core/Compilation/CompilationDifference.cs:153]: nit: spelling

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/EmitHelpers.vb [hunk 9]
```diff
@@ -34,6 +34,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
             Dim serializationProperties = compilation.ConstructModuleSerializationProperties(emitOpts, runtimeMDVersion, baseline.ModuleVersionId)
             Dim manifestResources = SpecializedCollections.EmptyEnumerable(Of ResourceDescription)()
 
+            Dim updatedMethods = ArrayBuilder(Of MethodDefinitionHandle).GetInstance()
```
Comment (RikkiGibson) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/EmitHelpers.vb:37]: I was confused by this as these don't look used (aside from ToImmutableAndFree()). Is there more work coming later?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/EmitHelpers.vb (hunk 9 above)
Comment (davidwengier) [reply src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/EmitHelpers.vb:37]: They're passed to `SerializeToDeltaStreams` on line 86. On the off chance this comment was meant for another location, there is another spot that probably has these arrays and there are changes coming in future, but the API is only in -vs-deps.

File: src/Compilers/Test/Core/Compilation/CompilationDifference.cs (hunk 8 above)
Comment (davidwengier) [reply src/Compilers/Test/Core/Compilation/CompilationDifference.cs:153]: Thank you! Will fix, and merge

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/EmitHelpers.vb (hunk 9 above)
Comment (RikkiGibson) [reply src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/EmitHelpers.vb:37]: I think I failed to use one of the following correctly:

1. My eyes
2. My code review tool

😄

**Pull request merged**